adam/learner/pursuit.py

```diff
@@ -657,16 +599,655 @@ def compute_hypothesis_id(h: PerceptionGraphPattern) -> str:
+        pattern = hypothesis
+        hypothesis_pattern_common_subgraph = get_largest_matching_pattern(
+            pattern,
+            graph,
+            debug_callback=self._debug_callback,
+            graph_logger=self._graph_logger,
+            ontology=self._ontology,
+            matching_objects=True,
+        )
+        self.debug_counter += 1
+
+        leading_hypothesis_num_nodes = len(pattern)
+        num_nodes_matched = (
+            len(hypothesis_pattern_common_subgraph.copy_as_digraph().nodes)
+            if hypothesis_pattern_common_subgraph
+            else 0
+        )
+        return ObjectPursuitLearner.ObjectHypothesisPartialMatch(
+            hypothesis_pattern_common_subgraph,
+            num_nodes_matched=num_nodes_matched,
+            num_nodes_in_pattern=leading_hypothesis_num_nodes,
+        )
+
     def _find_identical_hypothesis(
         self,
         new_hypothesis: PerceptionGraphPattern,
         candidates: Iterable[PerceptionGraphPattern],
     ) -> Optional[PerceptionGraphPattern]:
+        for candidate in candidates:
+            if new_hypothesis.check_isomorphism(candidate):
+                return candidate
+        return None
+
+    def _are_isomorphic(
+        self, h: PerceptionGraphPattern, hypothesis: PerceptionGraphPattern
+    ) -> bool:
+        return h.check_isomorphism(hypothesis)
+
+    @staticmethod
+    def from_parameters(
+        params: Parameters, *, graph_logger: Optional[HypothesisGraphLogger] = None
+    ) -> "ObjectPursuitLearner":  # type: ignore
+        log_word_hypotheses_dir = params.optional_creatable_directory(
+            "log_word_hypotheses_dir"
+        )
+        if log_word_hypotheses_dir:
+            logging.info("Hypotheses will be logged to %s", log_word_hypotheses_dir)
+
+        rng = Random()
+        rng.seed(params.optional_integer("random_seed", default=0))
+
+        return ObjectPursuitLearner(
+            learning_factor=params.floating_point("learning_factor"),
+            graph_match_confirmation_threshold=params.floating_point(
+                "graph_match_confirmation_threshold"
+            ),
+            lexicon_entry_threshold=params.floating_point("lexicon_entry_threshold"),
+            smoothing_parameter=params.floating_point("smoothing_parameter"),
+            graph_logger=graph_logger,
+            log_word_hypotheses_to=log_word_hypotheses_dir,
+            rng=rng,
+            ontology=GAILA_PHASE_1_ONTOLOGY,
+        )
+
+
+class PrepositionPursuitLearner(
+    Generic[PerceptionT, LinguisticDescriptionT],
+    AbstractPursuitLearner[
+        PrepositionSurfaceTemplate,
+        PrepositionPattern,
+        PerceptionT,
+        LinguisticDescriptionT,
+    ],
+):
+    """
+    An implementation of pursuit learner for preposition leaning
+    """
+
+    # Variables for tracking preposition phrase information. These are filled in observe.
+    object_match_node_for_ground: Optional[MatchedObjectNode] = None
```

Comment:
Yes, but they are used across different methods, both `observe()` and `candidate_perceptions()`

Reply:
hmm - that generally indicates we need a restructuring to avoid this, but let's not let it block the merge.  For now, note it on the other issue about further refactoring.